include/swift/SIL/MemoryLifetime.h

```diff
@@ -0,0 +1,374 @@
+
+  /// Helper function called by analyzeLocation to check all uses of the
+  /// location recursively.
+  ///
+  /// The \p subLocationMap is a temporary cache to speed up sub-location lookup.
+  bool analyzeLocationUsesRecursively(SILValue V, unsigned locIdx,
+                                      SmallVectorImpl<SILValue> &collectedVals,
+                                      SubLocationMap &subLocationMap);
+
+  /// Helper function called by analyzeLocation to create a sub-location for
+  /// and address projection and check all of its uses.
+  bool analyzeAddrProjection(
+      SILValue projection, unsigned parentLocIdx,unsigned fieldNr,
+      SmallVectorImpl<SILValue> &collectedVals, SubLocationMap &subLocationMap);
+
+  /// Calculates Location::numFieldsNotCoveredBySubfields
+  void initFieldsCounter(Location &loc);
+
+  /// Handles locations of the predecessor's terminator, which are only valid
+  /// in \p block.
+  /// Example: @out results of try_apply. They are only valid in the
+  /// normal-block, but not in the throw-block.
+  void setBitsOfPredecessor(Bits &bits, SILBasicBlock *block);
+
+  /// Sets up the BlockState datastructures and associates all basic blocks with
+  /// a state.
+  void initBlockStates();
+
+  /// Initializes the data flow bits sets in the block states for all blocks.
+  void initDataflow();
+
+  /// Initializes the data flow bits sets in the block state for a single block.
+  void initDataflowInBlock(BlockState &state);
+
+  /// Helper function to set bits for function arguments and returns.
+  void setFuncOperandBits(BlockState &state, Operand &op,
+                          SILArgumentConvention convention,
+                          bool isTryApply);
+
+  /// Calculates the BlockState::reachableFromEntry flags.
+  void entryReachabilityAnalysis();
+
+  /// Calculates the BlockState::exitReachable flags.
+  void exitReachableAnalysis();
+
+  /// Derives the block exit sets from the entry sets by applying the gen and
+  /// kill sets.
+  void solveDataflowForward();
+
+  /// Derives the block entry sets from the exit sets by applying the gen and
+  /// kill sets.
+  void solveDataflowBackward();
+
+  /// Do a block-local processing for all locations in singleBlockLocations.
+  ///
+  /// First, initializes all locations which are alive in a block and then
+  /// calls \p handlerFunc for the block.
+  void handleSingleBlockLocations(
+                       std::function<void (SILBasicBlock *block)> handlerFunc);
+
+  MemoryLifetime(SILFunction *F) : F(F) { }
+
+public:
+
+  /// Debug dump the MemoryLifetime internals.
+  void dump() const;
+
+  /// Debug dump a bit set .
+  static void dumpBits(const Bits &bits);
+
+};
+
+/// A utility for verifying memory lifetime.
+///
+/// The MemoryLifetime utility checks the lifetime of memory locations.
+/// This is limited to memory locations which are guaranteed to be not aliased,
+/// like @in or @inout parameters. Also, alloc_stack locations are handled.
```

Comment:
Shouldn't this also apply to things that have an appropriate begin_access, end_access so we know that we have unique access to them?

Reply:
Interesting idea. That could add some power to the verifier.